feat: support react-native-vision-camera v3

rkmackinnon opened the pull request:
react-native-vision-camera is currently undergoing a rewrite to a new major version, v3: [https://github.com/mrousavy/react-native-vision-camera/issues/1376](https://github.com/mrousavy/react-native-vision-camera/issues/1376). This pull request would update vision-camera-code-scanner to support it. This code was tested against commit [af4e366](https://github.com/mrousavy/react-native-vision-camera#af4e366312a2335535f80109e6bf0b3a977444f0) from March 21, 2023 in the v3 branch of react-native-vision-camera.

rocket13011 commented:
Hello,

Be careful a self in an override, I'm not sure swift likes that.

just replace self by the class name ;)

rkmackinnon commented:
Hi @rocket13011, thanks for the tip. I have not done too much Swift programming, so I welcome any advice.

It seems like self in this situation is just specifying that the code calls other methods on the same instance of the VisionCameraCodeScanner class. Is there something more complicated happening there?

React-native-vision-camera v3's approach no longer expects a static callback method in frame processor plugins, so I removed the static keywords from this class. Replacing self by the class name wouldn't seem to work anymore with that change, but if it is more idiomatic Swift to avoid using self unnecessarily then just removing it altogether is an option.

bglgwyng commented:
@rkmackinnon I tried installing this branch and got the following error message.
```
/asdf/node_modules/@bglgwyng/vision-camera-code-scanner/ios/VisionCameraCodeScanner-Bridging-Header.h:1:9: error: include of non-modular header inside framework module 'VisionCameraCodeScanner.VisionCameraCodeScanner_Bridging_Header': '/asdf/ios/Pods/Headers/Public/VisionCamera/FrameProcessorPlugin.h'
#import <VisionCamera/FrameProcessorPlugin.h>
```
I published the build of this branch to my private npm registry and installed it from the registry.
I suppose the 'example` in this repo would work, but have you tested it with a new project?

bglgwyng commented:
I found the same error occurs using the local zipped package as you suggested. 
I suppose that the error is related to my use of `use_frameworks! :linkage => :static` in `Podfile`. The project without that option is built well with this branch. However, the old 'vision-camera-code-scanner' worked fine with `use_frameworks! :linkage => :static`, so I think we need to figure out which change you made caused this difference.

Do you have any idea?

rkmackinnon commented:
@bglgwyng I suspect you're right. I am using Flipper so I do not have `use_frameworks! :linkage => :static` in my Podfile. It looks like react-native-skia did not add support for `use_frameworks! :linkage => :static` until https://github.com/Shopify/react-native-skia/commit/96168570e0ff1594e57419cbadb5a7bab528e5ee, which the oldest supported release is 0.1.176. I based this pull request off of 0.1.175 due to this other issue which to my knowledge hasn't been fixed yet: https://github.com/mrousavy/react-native-vision-camera/pull/1466#issuecomment-1460395999.

Regardless, I tried to bump the version of react-native-skia to the latest on my branch in case it contained the fix, enabled the static frameworks option in the example project and unfortunately ran into a bunch of other compilation errors. Despite ostensibly having the new frameworks compatible version of react-native-skia, I was seeing errors akin to https://github.com/Shopify/react-native-skia/issues/652#issuecomment-1208569391, where if I removed the path prefix I could get a bit further along in the compilation. That seems to be the wrong way to go about fixing the problem though.

If you were able to get a message specific to vision-camera-code-scanner, you must be doing something differently since skia doesn't seem to be a problem for you. What customizations did you make to the repo after cloning it to get your error message?

rocket13011 commented:
For skia patch :

`diff --git a/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/DerivedNodeProp.h b/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/DerivedNodeProp.h
index 64b817e..3a9ecb4 100644
--- a/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/DerivedNodeProp.h
+++ b/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/DerivedNodeProp.h
@@ -72,7 +72,7 @@ public:
    Adds a property to the derived property child props.
    */
   template <class _Tp, class... _Args,
-            class = std::_EnableIf<!std::is_array<_Tp>::value>>
+            class = std::enable_if_t<!std::is_array<_Tp>::value>>
   _Tp *defineProperty(_Args &&...__args) {
     auto prop =
         std::make_shared<_Tp>(std::forward<_Args>(__args)..., _onChange);
diff --git a/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/NodePropsContainer.h b/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/NodePropsContainer.h
index 380764c..10f1ac9 100644
--- a/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/NodePropsContainer.h
+++ b/node_modules/@shopify/react-native-skia/cpp/rnskia/dom/base/NodePropsContainer.h
@@ -106,7 +106,7 @@ public:
    Defines a property that will be added to the container
    */
   template <class _Tp, class... _Args,
-            class = std::_EnableIf<!std::is_array<_Tp>::value>>
+            class = std::enable_if_t<!std::is_array<_Tp>::value>>
   _Tp *defineProperty(_Args &&...__args) {
     // Create property and set onChange callback
     auto prop = `

bglgwyng commented:
The patch @rocket13011 provides was already applied to the project. I think that's why I didn't see the error related to react-native-skia.
[Here](https://github.com/bglgwyng/RNVC3Sandbox/tree/vision-camera-code-scanner-test) I have made a reproducible scenario. Unfortunately, I am unable to determine why the error message is slightly different, and it appears as follows:
```
/Users/sojaegyeong/Documents/GitHub/RNVC3Sandbox/node_modules/vision-camera-code-scanner/ios/VisionCameraCodeScanner-Bridging-Header.h:1:9: error: 'VisionCamera/FrameProcessorPlugin.h' file not found
#import <VisionCamera/FrameProcessorPlugin.h>
        ^
<unknown>:0: error: could not build Objective-C module 'VisionCameraCodeScanner'
```

Regardless, it is still true that the use of use_frameworks! :linkage => :static is the determining factor in the successful build.

@rkmackinnon Could you please test this repo and tell me if you can see the same error message?
I'll let you know as soon as I succeed at reproducing the exact same error message I reported yesterday.

rkmackinnon commented:
Hi again @bglgwyng. I was able to fix the error you were getting. There is a missing dependency on the VisionCamera library in the podspec file for VisionCameraCodeScanner. I will push a new commit to my branch to fix this. Unfortunately, the code still doesn't compile though. Now your project and mine are both experiencing the same errors I was talking about with skia. Maybe you'll have some more luck than I have been having fixing that, but I have not given up on it yet!

bglgwyng commented:
@rkmackinnon, thank you very much for your prompt response! I gave your latest commit a try, and the first time I ran it, I noticed that the error I previously reported was no longer present. However, it was replaced with a new issue related to 'react-native-skia'. Luckily, I was already familiar with how to resolve this type of problem, and I found a solution in this link: https://github.com/Shopify/react-native-skia/issues/652#issuecomment-1280278930.

Nevertheless, the strange thing is that even after I applied the fix, I encountered the 'error: could not build Objective-C module 'VisionCameraCodeScanner' again. You can observe this behavior in the following commit: https://github.com/bglgwyng/RNVC3Sandbox/commit/4c65b2fff9edef969d56d4aa0a0fd30d60a31d99.

rkmackinnon commented:
@bglgwyng It has been pleasant going back and forth with you on this as well. I believe I have fixed the latest issue you reported. If you pull from my branch again and re-run `npm pack` the package should be able to be used with `use_frameworks! :linkage => :static` or without it.

I added a patch to the react-native-vision-camera library that essentially added the an empty VisionCamera.h file to satisfy its complaints. I'm not sure if that's the best way to solve the problem, but it works! I tested it in the example project with `use_frameworks! :linkage => :static` as well as in my private repo that doesn't currently have that option.

bglgwyng commented:
@rkmackinnon, I followed your instructions as directed, but unfortunately, it did not work. [Here](https://github.com/bglgwyng/RNVC3Sandbox/commit/bf834713cb6e3a23e905ce4ef8579c3dc5e0daa8) is the commit that I have included the same patch file. Can you please tell me what I may have done incorrectly?

Furthermore, it would be nice to report the 'VisionCamera.h' issue on the V3 issue tracker (https://github.com/mrousavy/react-native-vision-camera/issues/1376), since it appears to be a general problem across all frame processor modules.

rkmackinnon commented:
Hi @bglgwyng, I took a look at your patch. I see you have patched the podspec file to include the reference to VisionCamera.h, but it looks like you maybe forgot to create the empty VisionCamera.h file in the patch as well? My patch from https://github.com/rodgomesc/vision-camera-code-scanner/blob/4dcfeabd3cd41564257e3ec1b0824f3a39b15014/.yarn/patches/react-native-vision-camera-https-95cda75d0b.patch also includes this part:

```
diff --git a/VisionCamera.h b/VisionCamera.h
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
```

That is a good suggestion about commenting on the VisionCamera.h file on the V3 issue tracker. There is actually an older issue https://github.com/mrousavy/react-native-vision-camera/issues/1043 about it that I could comment on as well, but I think I should probably do both as that older one might not be actively monitored anymore.

bglgwyng commented:
Your fix works. Thank you!

AnomalousLLC commented:
For Expo users who don't want to prebuild or eject: https://github.com/AnomalousLLC/vision-camera-code-scanner/commit/5d372922af7702cbc235020d3268646e0ec9e48f

AnomalousLLC commented:
Register plugin(s) outside of AppDelegate: https://github.com/AnomalousLLC/vision-camera-code-scanner/commit/5d372922af7702cbc235020d3268646e0ec9e48f

No need to register Swift plugins manually after installation, maintains out of the box Expo support without prebuild/ejecting.

lBroth commented:
Hello everyone, are there any updates on this PR? The vision camera v3 has just been released.

OnurGuersoy commented:
> Hi @OnurGuersoy
> 
> This PR does need some more work. There is a fix already merged to main in react-native-vision-camera since the latest v3.0.0 release that is required for this plugin to work properly on Android. I am just waiting for a new npm release that contains the fix to make the required updates for Android.
> 
> I will post a message in here when I think the PR is ready.

Cool thank you!

deeeed commented:
Looking forward to a fix

KristianSelnas commented:
Has anyone looked at this recently? react-native-vision-camera v3.1 is released now. Support for v3.x would be much appreciated.

RikSchefferOberon commented:
can someone publish this fix, so we can use v3?

rkmackinnon commented:
@RikSchefferOberon, the previous fix for Android was incomplete, so it would be inappropriate to publish it.

@KristianSelnas Frame processors still do not work on Android in the latest v3.1.0. This time we are potentially waiting on new PRs: https://github.com/mrousavy/react-native-vision-camera/pull/1873, https://github.com/mrousavy/react-native-vision-camera/pull/1874, perhaps others, and a new published version of react-native-vision-camera that includes them. See also: https://github.com/mrousavy/react-native-vision-camera/issues/1837, specifically the point about being restricted to rgb pixel format on Android and the implications that may have on the results returned from MLKit.

When a new version of react-native-vision-camera is released I would appreciate another head's up in case I miss it. In the meantime, please be patient.

KristianSelnas commented:
> Does anyone know another library I can use for using a camera and reading QR? I need to implement this feature in my project ASAP. Since we don't know how many days this fix will take to get published, I need a workaround.

https://github.com/teslamotors/react-native-camera-kit is an option, but personally I'm a bit dubious about the quality and maintenance of that library.

joacub commented:
> @RikSchefferOberon, the previous fix for Android was incomplete, so it would be inappropriate to publish it.
> 
> @KristianSelnas Frame processors still do not work on Android in the latest v3.1.0. This time we are potentially waiting on new PRs: [mrousavy/react-native-vision-camera#1873](https://github.com/mrousavy/react-native-vision-camera/pull/1873), [mrousavy/react-native-vision-camera#1874](https://github.com/mrousavy/react-native-vision-camera/pull/1874), perhaps others, and a new published version of react-native-vision-camera that includes them. See also: [mrousavy/react-native-vision-camera#1837](https://github.com/mrousavy/react-native-vision-camera/issues/1837), specifically the point about being restricted to rgb pixel format on Android and the implications that may have on the results returned from MLKit.
> 
> When a new version of react-native-vision-camera is released I would appreciate another head's up in case I miss it. In the meantime, please be patient.

vision camera already publish antoher version and frameprocessor works now in android

rkmackinnon commented:
@joacub
 
> ... and frameprocessor works now in android

Are you sure? When I tried it it didn't compile for Android.

joacub commented:
> @joacub
> 
> > ... and frameprocessor works now in android
> 
> Are you sure? When I tried it it didn't compile for Android.

Vision camera is working in android and iOS perfectly no problems

joacub commented:
> > @joacub
> > > ... and frameprocessor works now in android
> > 
> > 
> > Are you sure? When I tried it it didn't compile for Android.
> 
> Vision camera is working in android and iOS perfectly no problems I have been using since version 3.0 with no issues.

rkmackinnon commented:
Certainly some frame processors would've worked with version 3.0, but not this one. There was a bug in version 3.0 that caused Android frame processors that had options with array properties to crash. In version 3.1 that bug was fixed, but the new changes to add back in the OpenGL processing pipeline included new bugs that caused all Android frame processors to crash again. In version 3.2 those bugs seem to have been partially fixed.

You are welcome to submit a PR to my repo (the underlying repo that this PR is based on) if I am wrong about this.

rkmackinnon commented:
Please try this out. The example app should be working with both iOS and Android now.

azuddin commented on ios/VisionCameraCodeScanner.m:
```diff
@@ -1,5 +1,19 @@
+#if defined __has_include && __has_include("VisionCameraCodeScanner-Swift.h")
+#import "VisionCameraCodeScanner-Swift.h"
+#else
+#import <VisionCameraCodeScanner/VisionCameraCodeScanner-Swift.h>
```
where's this -Swift.h file located?

rkmackinnon replied:
It's an auto-generated file. The location is slightly different depending on whether the project is built with `use_frameworks! :linkage => :static`.

<img width="770" alt="image" src="https://github.com/rodgomesc/vision-camera-code-scanner/assets/129107344/14271802-756c-4ac7-8679-e00567c17785">

azuddin replied:
another thing, i did install this version. but whenever open the camera with frameProcessor, it crashes. any idea?

rkmackinnon replied:
Can you provide any more details of the crash (e.g., logs, exception messages) or your test environment?

mare95 replied:
I experienced the same issue. Camera works great but when i attach `frameProcessor` it crashes the app. This is the log i get from the xCode. Do you know what could be a solution for this issue?

```
VisionCamera.didSetProps(_:): Updating 16 prop(s)...
VisionCamera.configureCaptureSession(): Configuring Session...
VisionCamera.configureCaptureSession(): Initializing Camera with device com.apple.avfoundation.avcapturedevice.built-in_video:6...
VisionCamera.configureCaptureSession(): Adding Video input...
VisionCamera.configureCaptureSession(): Adding Video Data output...
VisionCamera.invokeOnInitialized(): Camera initialized!
VisionCamera.configureCaptureSession(): Session successfully configured!
VisionCamera.configureFormat(): Configuring Format...
VisionCamera.configureFormat(): Format successfully configured!
VisionCamera.configureDevice(): Configuring Device...
VisionCamera.configureDevice(): Device successfully configured!
VisionCamera.didSetProps(_:): Starting Session...
VisionCamera.didSetProps(_:): Started Session!
libc++abi: terminating due to uncaught exception of type facebook::jsi::JSINativeException: Compiling JS failed: 1:1:invalid empty parentheses '( )' Buffer size 3 starts with: 280a29
```
Btw thanks @rkmackinnon for the effort!

chernandezq commented:
How can I test this PR?

rkmackinnon commented:
> How can I test this PR @rkmackinnon ?

You can run the example app in the repo this PR is based on: https://github.com/rkmackinnon/vision-camera-code-scanner/tree/react-native-vision-camera-v3

superR8014 commented:
Hi @rkmackinnon
I try to Install with your latest commit, but shows below error
`$ bob build
ℹ Building target commonjs
ℹ Cleaning up previous build at lib/commonjs
ℹ Compiling 2 files in src with babel
✔ Wrote files to lib/commonjs
ℹ Building target module
ℹ Cleaning up previous build at lib/module
ℹ Compiling 2 files in src with babel
✔ Wrote files to lib/module
ℹ Building target typescript
ℹ Cleaning up previous build at lib/typescript
ℹ Generating type definitions with tsc
⚠ Failed to locate 'tsc' in the workspace. Falling back to the globally installed version. Consider adding typescript to your devDependencies or specifying the tsc option for the typescript target.
example/src/Constants.ts:2:34 - error TS2307: Cannot find module 'react-native-static-safe-area-insets' or its corresponding type declarations.

2 import StaticSafeAreaInsets from 'react-native-static-safe-area-insets';
                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in example/src/Constants.ts:2`

I have already tried to add 'react-native-static-safe-area-insets' this package but not useful.
Is there any suggestions?
Thanks!

rkmackinnon commented:
Hmm, I am using yarn to develop this package so it might be there are some tricky bits with working from source with npm. Try running yarn once you're in the example directory instead.

superR8014 commented:
> Hmm, I am using yarn to develop this package so it might be there are some tricky bits with working from source with npm. Try running yarn once you're in the example directory instead.

Thanks for your reply. Maybe I describe my problem not so clear.

I tried to download the package run in local with your suggestion "run yarn in the example directory" could run the project successfully.

But my current issue is I want to use this package to replace the original v0.2.0 vision-camera-code-scanner in my current project. So I just `yarn add https://github.com/rkmackinnon/vision-camera-code-scanner.git#react-native-vision-camera-v3` in my project but it gives me below error.
`$ bob build
ℹ Building target commonjs
ℹ Cleaning up previous build at lib/commonjs
ℹ Compiling 2 files in src with babel
✔ Wrote files to lib/commonjs
ℹ Building target module
ℹ Cleaning up previous build at lib/module
ℹ Compiling 2 files in src with babel
✔ Wrote files to lib/module
ℹ Building target typescript
ℹ Cleaning up previous build at lib/typescript
ℹ Generating type definitions with tsc
⚠ Failed to locate 'tsc' in the workspace. Falling back to the globally installed version. Consider adding typescript to your devDependencies or specifying the tsc option for the typescript target.
example/src/Constants.ts:2:34 - error TS2307: Cannot find module 'react-native-static-safe-area-insets' or its corresponding type declarations.

2 import StaticSafeAreaInsets from 'react-native-static-safe-area-insets';
                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in example/src/Constants.ts:2

`
I currently using react-native 0.72.4. Your help will be greatly appreciated.

hssrrw commented:
I'm experiencing the same problem. Somehow a new dependency in the examples (react-native-static-safe-area-insets) requires it to be installed, but we don't use it.

A possible solution would be ether (1) somehow ignore the dependencies from examples folder or (2) change the example to avoid the new dependency.

hssrrw commented:
Locally cloning and switching to this branch gives the same error when trying `npm install`.
I confirm, `yarn` works. However, we don't use yarn in our project. Looks like currently there is some incompatibility with npm cli but I am not sure what is it exactly.

hssrrw commented:
@rkmackinnon I know, Dimensions API might not be reliable on Android. That's probably why react-native-static-safe-area-insets was added in the first place. However, would that be sufficient to use just Dimensions in the example? Maybe we don't need precise window size for the demo purpose? I tested without this dependency and the package is installed in our project without any problem with npm.

You can check the changes here:
https://github.com/hssrrw/vision-camera-code-scanner/blob/4fb4a66340fdf922d6b859a617d10323fe171aec/example/src/Constants.ts

mare95 commented:
> After spending hours on debugging and understanding why my app crashes i figured that it's related to 'react-native-reanimated'.
> 
> This is the error: `libc++abi: terminating due to uncaught exception of type facebook::jsi::JSINativeException: Compiling JS failed: 1:1:invalid empty parentheses '( )' Buffer size 3 starts with: 280a29`
> 
> I'm using reanimated package for all different kind of animations and it can't be removed. These packages used to work well in V2.
> 
> Example app works but when i install `react-native-reanimated: "^3.5.4"` and when i add `['react-native-reanimated/plugin']` in `babel.config.js` => app crashes when there is a frameProcessor attached to the `<Camera />` component.
> 
> Steps to reproduce this:
> 
> * Clone project
> * Checkout to `vision-camera-code-scanner-v3`
> * cd into example project
> * Add `"react-native-reanimated": "^3.5.4",` package
> * open babel.config.js and update it like this
> 
> ```
> const pak = require('../package.json');
> 
> module.exports = {
>   presets: ['module:metro-react-native-babel-preset'],
>   plugins: [
>     [
>       'module-resolver',
>       {
>         extensions: ['.tsx', '.ts', '.js', '.json'],
>         alias: {
>           [pak.name]: path.join(__dirname, '..', pak.source),
>         },
>       },
>     ],
>     ['react-native-worklets-core/plugin'],
>     ['react-native-reanimated/plugin'],
>   ],
> };
> ```
> 
> * Install dependencies and pods
> * Open example/ios in xCode (Version 15.0 (15A240d))
> * Build the app
>   => When you scan the code app will crash (see attached image)
> 
> Do you know what could be the reason and why `vision-camera-code-scanner` and `react-native-vision-camera` don't work with `react-native-reanimated` anymore?
> 
> Btw this all worked with v2 but i had to update it for iOS 17 and iPhone 15 <img alt="Screenshot 2023-10-03 at 17 12 26" width="1440" src="https://user-images.githubusercontent.com/17355348/272318766-273f1073-edb0-43e4-b7b1-6c405d6fd92c.png">

Solution was to bump "react-native-worklets-core": "^0.2.0" to "react-native-worklets-core": "^0.2.1"

rkmackinnon commented:
@superR8014 @hssrrw I believe this latest commit will fix the issue with react-native-statis-safe-area-insets while still allowing it to be used in the example project. Thanks for helping test this.

rkmackinnon commented:
@mare95 Thanks for all your time on this, it is much appreciated. We do not use reanimated so had not run into this incompatibility ourselves. I will bump the react-native-worklets-core version on this branch.

xhoantran commented:
Verified this work on IOS. Thanks @rkmackinnon

mare95 commented:
Yup, my app works in "dev" environment but as soon as i build it for test it crashes when i open the camera that has `frameProcessor` attached with this error:

```
libc++abi: terminating due to uncaught exception of type facebook::jsi::JSINativeException: Compiling JS failed: 1:1:invalid empty parentheses '( )' Buffer size 3 starts with: 280a29
```

Here is the screenshoot from firebase crashlytics as well:
<img width="1005" alt="Screenshot 2023-10-04 at 14 50 31" src="https://github.com/rodgomesc/vision-camera-code-scanner/assets/17355348/1aa76210-e7e5-4185-8865-af7a36b00a47">

Any ideas why?

vietdung97 commented:
> For anyone seeing these crashes, can you reproduce it in the example app?
example app run perfectly, but our app still get crash issue.
my package.json:
"react-native-vision-camera": "^3.2.2",
"react-native-worklets-core": "^0.2.1",
"vision-camera-code-scanner": "https://github.com/rkmackinnon/vision-camera-code-scanner.git#react-native-vision-camera-v3",

xhoantran commented:
"react-native-vision-camera" now support code scanner 
https://github.com/mrousavy/react-native-vision-camera/releases/tag/v3.3.0

vietdung97 commented:
> "react-native-vision-camera" now support code scanner https://github.com/mrousavy/react-native-vision-camera/releases/tag/v3.3.0

Thank bro, you save my day

rkmackinnon commented:
> "react-native-vision-camera" now support code scanner https://github.com/mrousavy/react-native-vision-camera/releases/tag/v3.3.0

Always great to have another option, especially one tied to a project with a better history of ongoing maintenance/support. I'll definitely be trying that out. At first glance, because it's not a frame processor, it looks like it'll support older devices on the Android side. One advantage of vision-camera-code-scanner is that we expose more of the structured output of MLKit. We also have the option to check inverted colors. It'll be interesting to see if there's enough interest to keep this project going after everyone has the time to make some comparisons and consider what they need.

zzz08900 commented:
On iOS side the built-in code scanner of vision camera v3 uses some iOS built-in function to read QR codes and by the result of initial testing I personally prefer the results from MLKit.
